src/OrbitBase/include/OrbitBase/Overloaded.h

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_BASE_OVERLOADED_H_
+#define ORBIT_BASE_OVERLOADED_H_
+
+#include <functional>
+#include <iostream>
+#include <type_traits>
+#include <utility>
+
+namespace orbit_base_internal {
+
+// The class wraps a function pointer, providing `operator()`
+template <typename FunctionPtr>
+class FunctionPtrWithParen {
+ public:
+  FunctionPtrWithParen(FunctionPtr t) : function_ptr_(t) {}  // NOLINT(google-explicit-constructor)
+
+  template <typename... In>
+  auto operator()(In&&... in) const
+      -> decltype(std::invoke(std::declval<FunctionPtr>(), std::forward<In>(in)...)) {
+    return std::invoke(function_ptr_, std::forward<In>(in)...);
+  }
+
+ private:
+  FunctionPtr function_ptr_;
+};
+
+// Compile-time check if `operator()` is defined for `T`
+template <typename T, typename = std::void_t<>>
+struct HasOperatorParen : std::false_type {};
+
+template <typename T>
+struct HasOperatorParen<T, std::void_t<decltype(&T::operator())>> : std::true_type {};
+
+template <typename T>
+inline constexpr bool kHasParenV = HasOperatorParen<T>::value;
+
+// If `T` has `operator()` defined, is `T`. Otherwise it is `FunctionPtrWrapper<T>`
```

Comment:
", is `T`. " -> ", *it* is `T`. "

Or maybe: `T` if `operator()` is defined, otherwise `FunctionPtrWrapper<T>`